Add cli `key child` command

Anviking opened the pull request:
# Issue Number

#1353 

# Overview

- [x] Preparatory addition of CLI `DerivationIndex` and `DerivationPath` types
- [x] Make `mapKey` xprv to hex transformation bidirectional
- [x] Add `key child --path` command to cli
- [x] Confirming that we can derive an actual wallet address key using `key child` as a manual golden test.

# Comments
- [ ] Not here: Parsing hexadecimal indexes e.g `0x8000073c` (could be neat 🤷‍♂ )

Manual example. Note that `to-pub` is not functionality of `cardano-wallet` — yet.
```bash
$ cardano-wallet-jormungandr key child --path "2147485500/2147485463/0'/0/0" 588102383ed9ecc5c44e1bfa18d1cf8ef19a7cf806a20bb4cbbe4e511666cf48d6fd7bec908e4c6ced5f0c4f0798b1b619d6b61e6110492b5ebb430f570488f074a9fc9a22f0a61b2ab9b1f1a990e3f8dd6fbed4ad474371095c74db3d9c743a | to-pub
eb806fca3bfda11ae14dee52c5ad6564e68e7531c1a1d498a76172898d2841cb3d9078e802011f1580465c80e7040f1e4d8e24f978d23f01c1d2cf18fcf741a7

# Address comes from an actual empty wallet:
$ jcli address info addr1sn4cqm728076zxhpfhh993ddv4jwdrn4x8q6r4yc5ash9zvd9pqukyh6fnkrsrsz4w9jvu8hyre20xm7m78jf4kww4h2n5maytuhtflrax9duz
discrimination: testing
public key: ed25519_pk1awqxlj3mlks34c2daefvttt9vnnguaf3cxsafx98v9egnrfgg89s4y7wfm
group key:  ed25519_pk1ztayempcpcp2hzexwrmjpu48ndldlrey6m882m4f6d7j97t45l3s5h5gl4
```

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -431,6 +447,64 @@ instance FromText CliWalletStyle where
+
+        parseHardenedIndex =
+            parseWord
+            >=> (return . (+ 0x80000000))
```
Should probably leverage the existing type
```Haskell
newtype Index (derivationType :: DerivationType) (level :: Depth) = Index
```
to avoid hard-coding `0x80000000` again.

Anviking replied:
However, I'm not convinced that we need as much flexibility as the `Index` type provides.

For now I resorted to defining
```Haskell
firstHardenedIndex :: Word32
firstHardenedIndex = getIndex $ minBound @(Index 'Hardened 'AddressK)
```

Anviking committed:
Preparatory addition of CLI `DerivationIndex` and `DerivationPath` types

Anviking commented on lib/cli/test/unit/Cardano/CLISpec.hs:
```diff
@@ -455,22 +458,63 @@ spec = do
+        textRoundtrip $ Proxy @DerivationPath
+
+    describe "DerivationPath/Index fromText goldens" $ do
+        let golden str desc expected = it (show str ++ " " ++ desc) $ fromText str `shouldBe` expected
```
A few lines here are too long. Should look over.

Anviking commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -501,9 +594,33 @@ cmdRootKey =
+        scheme =
+            hoistKeyScheme eitherToIO
+            . mapKey xPrvToTextTransform
+            $ newCliKeyScheme style
```
Ok, I see 

> No, rather a --byron flag, if present, use derivation scheme 1, if not, use derivation scheme 2.

that we shouldn't take `--wallet-style` as an argument.

`CliKeyScheme` might need to be broken apart then. 🤔

Anviking replied:
I guess one _could_ strategically use a hard-coded `Icarus` scheme here and leave a todo for potential second U/S. And add three goldens for `key child` such that they don't end up diverging by accident.

Anviking replied:
>  I guess one could strategically use a hard-coded `Icarus`

Done.

Anviking committed:
Rename cmdRootKey -> cmdKeyRoot

paweljakubas commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -378,44 +386,54 @@ hoistKeyScheme
-xPrvToHexText = left showErr . fmap (T.pack . B8.unpack . hex) . unXPrvStripPub
+-- | Pair of functions representing the bidirectional transformation between
+-- a @XPrv@ and its 96-byte hex-encoded form.
+xPrvToTextTransform :: (XPrv -> Either String Text, Text -> Either String XPrv)
```
can't we just use `FromText` and `ToText` of `XPrv` (or probably some wrapped `XPrv` like we do in `Api.Types` using `ApiT` to avoid orphan instances) instances instead?

Anviking replied:
> can't we just use FromText and ToText of XPrv (or probably some wrapped XPrv like we do in

We _could_ probably have gone that route.

I thought this way played nicely with
```Haskell
data CliKeyScheme key m = CliKeyScheme
    { allowedWordLengths :: [Int]
    , mnemonicToRootKey :: [Text] -> m key
    , deriveChildKey :: key -> DerivationIndex -> m key
    }
```
since it
- Encapsulated more logic in `CliKeyScheme` that could be tested easily. It enables writing the `scheme == mapKey (fromHex . toHex) scheme` property.*
- More easily allows multiple encodings (e.g. if wanted to add additional support for bech32 at some point)

*) Although I'm not clear on exactly what value this brings compared two normal `FromText`/`ToText` roundtrip properties. Might (just) be to ensure the roundtrips are performed with the all XPrvs that are used for the other tests, and not a subset.

Had we supported `Byron`, it would have made sure DerivationScheme1 child keys also round-tripped, without us having to add DerivationScheme1 child-keys to the `Arbitrary XPrv` generator (not an obvious thing to do).

paweljakubas commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -378,44 +386,54 @@ hoistKeyScheme
+    xPrvToHexText = left showErr . fmap (T.pack . B8.unpack . hex) . unXPrvStripPub
+      where
+        showErr ErrCannotRoundtrip =
+            "That private key looks weird. Is it encrypted? Or an old Byron key?"
```
no chance to say precisely here whether it is old Byron key or the other?

Anviking replied:
Maybe.

The private part of a Byron key will be changed by the tweak. So both the private key and the public key will be different.

An encrypted `DerivationScheme2` key will roundtrip the private key correctly. But it will re-derive a different public key.

**However**, this error message is for the `XPrv -> Hex` conversion. It is only used to encode the resulting child key. It should actually never be called* 🙃 🤔 👀 

*) except if we were to re-introduce derivation using `DerivationScheme1` / Byron.

rvl commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -431,6 +449,79 @@ instance FromText CliWalletStyle where
+
+        parseHardenedIndex =
+            parseWord
+            >=> (return . (+ fromIntegral firstHardenedIndex))
```
I think do notation or a fmap would be clearer here.

Anviking replied:
I removed some `>=>` in a commit

rvl commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -1263,6 +1399,20 @@ walletStyleOption = option (eitherReader fromTextS)
+    <> value ([DerivationIndex 0x80000000])
+  where
+    f :: String -> Either String [DerivationIndex]
+    f s = left getTextDecodingError $ mapM fromText $ T.splitOn "/" (T.pack s)
```
Is the empty derivation path handled anywhere?

This seems to overlap the functionality of `instance FromText DerivationPath`.

Anviking replied:
> This seems to overlap the functionality o

😮 Indeed, thanks!

> Is the empty derivation path handled anywhere?

It is currently handled somewhat implicitly by `T.splitOn "/"` returning `""`, which then results in the `DerivationIndex` failing with _An empty string is not a derivation index!_

Perhaps it could be desirable to allow empty paths (returning the input key), not sure. In-doubt I thought it would probably be a user error, and the CLI erroring would be desirable.

rvl commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -431,6 +449,79 @@ instance FromText CliWalletStyle where
+        "An empty string is not a derivation index!"
+    fromText x = do
+       -- NOTE: T.takeEnd will not throw, but may return "".
+       if T.takeEnd 1 x == "\'"
```
I put a comment on ADP-81 about this. I don't think using the `'` character will work very well in the most common shells (sh, bash, zsh).

Anviking replied:
I think for instance the risk of turning `0'/0'` into `0/0` by accident makes this a valid concern. I changed to `H` as suggested by Matthias.

rvl reviewed (approved):
Looks good

Anviking commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -1263,6 +1399,20 @@ walletStyleOption = option (eitherReader fromTextS)
+    <> long "path"
+    <> metavar "DER-PATH"
+    <> help "Derivation path e.g. \"44'/1815'/0'/0\""
+    <> value ([DerivationIndex 0x80000000])
```
I think this default `value` _definitely_ shouldn't be here.

Anviking committed:
Strip whitespace when decoding XPrvs

otherwise $() composition won't work

Anviking committed:
Make icarus the default --wallet-style

And tiny tweak of test descriptions

Anviking committed:
Add `key child --path` command to cli

Tests extended to verify that key derivation is unaffected by hex
roundtrips.

Anviking committed:
Rename metavar HEX-XPRV -> XPRV

Anviking committed:
Replace 0' notation with 0H

This avoids accidents where two ' would be "removed" by the user's shell

Anviking committed:
Add some extra golden tests for derivation paths

Anviking committed:
fixup: remove some use of (>=>)

Anviking committed:
Add xPrvFromStrippedPubXPrvCheckRoundtrip (and some related rework)

Checking roundtrips is actually more important in this direction. To
avoid infinite loops, I made sure we have non-roundtrip-checking
variants as well.

Now we have:
unXPrvStripPub
unXPrvStripPubCheckRoundtrip
xPrvFromStrippedPubXPrv
xPrvFromStrippedPubXPrvCheckRoundtrip

Anviking committed:
Test that cli doesn't accept byron keys

Anviking committed:
minor fixups in CLISpec

Anviking committed:
improve unXPrvStripPub comments and tests

paweljakubas commented on lib/cli/test/unit/Cardano/CLISpec.hs:
```diff
@@ -488,6 +511,40 @@ spec = do
+
+        fromTextGolden @DerivationPath "44H/0H/0" "" $
+            Right . DerivationPath . map DerivationIndex $
+                [0x80000000 + 44, 0x80000000 + 0, 0]
```
maybe we can give name to `0x8000000` - it is minBound of soft derivation index, right?

Anviking replied:
It's the minBound of the **hardened** indexes. Can move to a named let-binding just here

(wanted to duplicate the definition at https://github.com/input-output-hk/cardano-wallet/blob/c0479c0258bd6a8e8fd00d5e708a5fbe2e3fcfba/lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs#L230 on purpose though)

paweljakubas replied:
*hardened* yes yes

Anviking committed:
fixup: avoid magic 0x80000000 in spec

Anviking committed:
Acknowledge that CLI shouldn't accept encrypted keys

1. Add golden test for it
2. Make sure error messages mention it

Anviking committed:
Test CLI failure for XPRVs of wrong length
